File: agent/src/context/conversationContextManager.ts
import type { ConversationId, MessageId } from "../core/agentTypes.ts";
import { ZERO_ID, isZeroId } from "../core/agentTypes.ts";
import type {
  ConversationContextEntry,
  ConversationContextEntryStatus as ConversationContextEntryStatusType,
  ConversationContextState,
  ConversationPhase as ConversationPhaseType,
} from "./conversationContext.ts";
import {
  ConversationContextEntryKind,
  ConversationContextEntryStatus,
  ConversationPhase,
  rejectedOptionContent,
} from "./conversationContext.ts";

export const ConversationContextPatchOperation = {
  UpsertEntry: "upsert_entry",
  MarkEntryStatus: "mark_entry_status",
  RemoveEntry: "remove_entry",
  UpdateSummary: "update_summary",
  SetLastSummarizedMessage: "set_last_summarized_message",
  SetPhase: "set_phase",
} as const;

export type ConversationContextPatchOperation =
  (typeof ConversationContextPatchOperation)[keyof typeof ConversationContextPatchOperation];

export const ConversationContextPatchAuthor = {
  Owner: "owner",
  System: "system",
} as const;

export type ConversationContextPatchAuthor =
  (typeof ConversationContextPatchAuthor)[keyof typeof ConversationContextPatchAuthor];

export interface ConversationContextPatchItem {
  operation: ConversationContextPatchOperation;
  entry?: ConversationContextEntry;
  targetEntryId?: string;
  status?: ConversationContextEntryStatusType;
  summary?: string;
  messageId?: MessageId;
  phase?: ConversationPhaseType;
}

export interface ConversationContextPatch {
  conversationId: ConversationId;
  author: ConversationContextPatchAuthor;
  updatedAtMs: number;
  items: ConversationContextPatchItem[];
}

export interface ConversationContextApplyResult {
  ok: boolean;
  error: string;
  state: ConversationContextState;
}

export interface ConversationContextPolicy {
  maxUnsummarizedMessages: bigint;
}

export const DEFAULT_CONVERSATION_CONTEXT_POLICY: ConversationContextPolicy = {
  maxUnsummarizedMessages: 40n,
};

export class ConversationContextManager {
  readonly #policy: ConversationContextPolicy;

  constructor(policy: ConversationContextPolicy = DEFAULT_CONVERSATION_CONTEXT_POLICY) {
    this.#policy = policy;
  }

  applyPatch(state: ConversationContextState, patch: ConversationContextPatch): ConversationContextApplyResult {
    const next = structuredClone(state);

    if (isZeroId(patch.conversationId)) {
      return applyFailure("patch missing conversation_id", next);
    }
    if (!isZeroId(next.conversationId) && next.conversationId !== patch.conversationId) {
      return applyFailure("patch conversation_id does not match state", next);
    }
    if (isZeroId(next.conversationId)) {
      next.conversationId = patch.conversationId;
    }

    for (const item of patch.items) {
      switch (item.operation) {
        case ConversationContextPatchOperation.UpsertEntry: {
          if (item.entry && this.entryRequiresOwner(item.entry) && patch.author !== ConversationContextPatchAuthor.Owner) {
            return applyFailure("decision-zone entries require owner patch author", next);
          }
          if (!item.entry || !this.upsertEntry(next, item.entry, patch.updatedAtMs)) {
            return applyFailure("failed to upsert conversation context entry", next);
          }
          break;
        }
        case ConversationContextPatchOperation.MarkEntryStatus: {
          const target = item.targetEntryId ?? "";
          const status = item.status ?? ConversationContextEntryStatus.Active;
          if (this.targetEntryRequiresOwner(next, target) && patch.author !== ConversationContextPatchAuthor.Owner) {
            return applyFailure("decision-zone entry status changes require owner patch author", next);
          }
          if (!this.markEntryStatus(next, target, status, patch.updatedAtMs)) {
            return applyFailure(`conversation context entry not found: ${target}`, next);
          }
          break;
        }
        case ConversationContextPatchOperation.RemoveEntry: {
          const target = item.targetEntryId ?? "";
          if (this.targetEntryRequiresOwner(next, target) && patch.author !== ConversationContextPatchAuthor.Owner) {
            return applyFailure("decision-zone entry removal requires owner patch author", next);
          }
          if (!this.removeEntry(next, target)) {
            return applyFailure(`conversation context entry not found: ${target}`, next);
          }
          break;
        }
        case ConversationContextPatchOperation.UpdateSummary: {
          next.summary = item.summary ?? "";
          break;
        }
        case ConversationContextPatchOperation.SetLastSummarizedMessage: {
          next.lastSummarizedMessageId = item.messageId ?? ZERO_ID;
          break;
        }
        case ConversationContextPatchOperation.SetPhase: {
          if (patch.author !== ConversationContextPatchAuthor.Owner) {
            return applyFailure("conversation phase changes require owner patch author", next);
          }
          if (!item.phase) {
            return applyFailure("conversation phase patch missing phase", next);
          }
          next.phase = item.phase;
          break;
        }
      }
    }

    if (patch.updatedAtMs !== 0) {
      next.updatedAtMs = patch.updatedAtMs;
    }

    return applySuccess(next);
  }

  needsSummarization(state: ConversationContextState, latestMessageId: MessageId): boolean {
    if (isZeroId(latestMessageId) || BigInt(latestMessageId) <= BigInt(state.lastSummarizedMessageId)) {
      return false;
    }
    return BigInt(latestMessageId) - BigInt(state.lastSummarizedMessageId) >= this.#policy.maxUnsummarizedMessages;
  }

  private upsertEntry(state: ConversationContextState, incoming: ConversationContextEntry, updatedAtMs: number): boolean {
    const entry = structuredClone(incoming);
    if (!this.normalizeRejectedOptionEntry(entry)) {
      return false;
    }
    if (entry.content.length === 0) {
      return false;
    }
    if (entry.id.length === 0) {
      entry.id = this.nextEntryId(state);
    }
    if (updatedAtMs !== 0) {
      entry.updatedAtMs = updatedAtMs;
      if (entry.createdAtMs === 0) {
        entry.createdAtMs = updatedAtMs;
      }
    }

    const index = state.entries.findIndex((existing) => existing.id === entry.id);
    if (index === -1) {
      state.entries.push(entry);
      return true;
    }

    const existing = state.entries[index];
    if (existing && entry.createdAtMs === 0) {
      entry.createdAtMs = existing.createdAtMs;
    }
    state.entries[index] = entry;
    return true;
  }

  private markEntryStatus(
    state: ConversationContextState,
    entryId: string,
    status: ConversationContextEntryStatusType,
    updatedAtMs: number,
  ): boolean {
    const entry = state.entries.find((candidate) => candidate.id === entryId);
    if (!entry) {
      return false;
    }
    entry.status = status;
    if (updatedAtMs !== 0) {
      entry.updatedAtMs = updatedAtMs;
    }
    return true;
  }

  private removeEntry(state: ConversationContextState, entryId: string): boolean {
    const before = state.entries.length;
    state.entries = state.entries.filter((entry) => entry.id !== entryId);
    return state.entries.length !== before;
  }

  private nextEntryId(state: ConversationContextState): string {
    return `ctx_${state.conversationId}_${state.entries.length + 1}`;
  }

  private entryRequiresOwner(entry: ConversationContextEntry): boolean {
    return entry.kind === ConversationContextEntryKind.Decision || entry.kind === ConversationContextEntryKind.RejectedOption;
  }

  private targetEntryRequiresOwner(state: ConversationContextState, entryId: string): boolean {
    const entry = state.entries.find((candidate) => candidate.id === entryId);
    return entry ? this.entryRequiresOwner(entry) : false;
  }

  private normalizeRejectedOptionEntry(entry: ConversationContextEntry): boolean {
    if (entry.kind !== ConversationContextEntryKind.RejectedOption) {
      return true;
    }

    const record = entry.rejectedOption;
    if (!record || record.option.length === 0 || record.reason.length === 0 || record.premise.length === 0) {
      return false;
    }

    if (entry.content.length === 0) {
      entry.content = rejectedOptionContent(record);
    }
    return true;
  }
}

function applySuccess(state: ConversationContextState): ConversationContextApplyResult {
  return { ok: true, error: "", state };
}

function applyFailure(error: string, state: ConversationContextState): ConversationContextApplyResult {
  return { ok: false, error, state };
}

File: agent/native-cpp/src/context/conversation_context_manager.cpp
#include "agent/context/conversation_context_manager.h"

#include <algorithm>
#include <sstream>
#include <utility>

namespace atrium::agent {

ConversationContextApplyResult ConversationContextApplyResult::success(ConversationContextState state) {
    return ConversationContextApplyResult{true, "", std::move(state)};
}

ConversationContextApplyResult ConversationContextApplyResult::failure(std::string error, ConversationContextState state) {
    return ConversationContextApplyResult{false, std::move(error), std::move(state)};
}

ConversationContextManager::ConversationContextManager(ConversationContextPolicy policy) : m_policy(policy) {}

ConversationContextApplyResult ConversationContextManager::applyPatch(ConversationContextState state, const ConversationContextPatch &patch) const {
    if (patch.conversation_id == 0) {
        return ConversationContextApplyResult::failure("patch missing conversation_id", std::move(state));
    }
    if (state.conversation_id != 0 && state.conversation_id != patch.conversation_id) {
        return ConversationContextApplyResult::failure("patch conversation_id does not match state", std::move(state));
    }
    if (state.conversation_id == 0) {
        state.conversation_id = patch.conversation_id;
    }

    for (const auto &item : patch.items) {
        switch (item.operation) {
            case ConversationContextPatchOperation::UpsertEntry: {
                if (!upsertEntry(state, item.entry, patch.updated_at_ms)) {
                    return ConversationContextApplyResult::failure("failed to upsert conversation context entry", std::move(state));
                }
                break;
            }
            case ConversationContextPatchOperation::MarkEntryStatus: {
                if (!markEntryStatus(state, item.target_entry_id, item.status, patch.updated_at_ms)) {
                    return ConversationContextApplyResult::failure("conversation context entry not found: " + item.target_entry_id, std::move(state));
                }
                break;
            }
            case ConversationContextPatchOperation::RemoveEntry: {
                if (!removeEntry(state, item.target_entry_id)) {
                    return ConversationContextApplyResult::failure("conversation context entry not found: " + item.target_entry_id, std::move(state));
                }
                break;
            }
            case ConversationContextPatchOperation::UpdateSummary: {
                state.summary = item.summary;
                break;
            }
            case ConversationContextPatchOperation::SetLastSummarizedMessage: {
                state.last_summarized_message_id = item.message_id;
                break;
            }
        }
    }

    if (patch.updated_at_ms != 0) {
        state.updated_at_ms = patch.updated_at_ms;
    }

    return ConversationContextApplyResult::success(std::move(state));
}

bool ConversationContextManager::needsSummarization(const ConversationContextState &state, MessageId latest_message_id) const {
    if (latest_message_id == 0 || latest_message_id <= state.last_summarized_message_id) {
        return false;
    }
    return latest_message_id - state.last_summarized_message_id >= m_policy.max_unsummarized_messages;
}

bool ConversationContextManager::upsertEntry(ConversationContextState &state, ConversationContextEntry entry, std::uint64_t updated_at_ms) const {
    if (entry.content.empty()) {
        return false;
    }
    if (entry.id.empty()) {
        entry.id = nextEntryId(state);
    }
    if (updated_at_ms != 0) {
        entry.updated_at_ms = updated_at_ms;
        if (entry.created_at_ms == 0) {
            entry.created_at_ms = updated_at_ms;
        }
    }

    auto it = std::find_if(state.entries.begin(), state.entries.end(), [&entry](const ConversationContextEntry &existing) {
        return existing.id == entry.id;
    });

    if (it == state.entries.end()) {
        state.entries.emplace_back(std::move(entry));
        return true;
    }

    if (entry.created_at_ms == 0) {
        entry.created_at_ms = it->created_at_ms;
    }
    *it = std::move(entry);
    return true;
}

bool ConversationContextManager::markEntryStatus(ConversationContextState &state, const std::string &entry_id, ConversationContextEntryStatus status, std::uint64_t updated_at_ms) const {
    auto it = std::find_if(state.entries.begin(), state.entries.end(), [&entry_id](const ConversationContextEntry &entry) {
        return entry.id == entry_id;
    });
    if (it == state.entries.end()) {
        return false;
    }
    it->status = status;
    if (updated_at_ms != 0) {
        it->updated_at_ms = updated_at_ms;
    }
    return true;
}

bool ConversationContextManager::removeEntry(ConversationContextState &state, const std::string &entry_id) const {
    auto before = state.entries.size();
    state.entries.erase(std::remove_if(state.entries.begin(), state.entries.end(), [&entry_id](const ConversationContextEntry &entry) {
        return entry.id == entry_id;
    }), state.entries.end());
    return state.entries.size() != before;
}

std::string ConversationContextManager::nextEntryId(const ConversationContextState &state) const {
    std::ostringstream out;
    out << "ctx_" << state.conversation_id << "_" << (state.entries.size() + 1);
    return out.str();
}

const char *toString(ConversationContextPatchOperation operation) {
    switch (operation) {
        case ConversationContextPatchOperation::UpsertEntry: return "upsert_entry";
        case ConversationContextPatchOperation::MarkEntryStatus: return "mark_entry_status";
        case ConversationContextPatchOperation::RemoveEntry: return "remove_entry";
        case ConversationContextPatchOperation::UpdateSummary: return "update_summary";
        case ConversationContextPatchOperation::SetLastSummarizedMessage: return "set_last_summarized_message";
    }
    return "unknown";
}

} // namespace atrium::agent

